tim-hm/ratings
# feat: sqlx migrations in test hooks

Changes:
- Created a new db user and connection for it with different permissions to manage the migrations.
- Migrations run when the first test is run and rolled back when the last has finished.
- Removed `INITIALIZATION_FLAG` from hooks. Was having some issues with threads yielding as strange times causing issues so Swapped to a `call_once()`
- Split the migrations into two folders, one for sqlx to manage and one that will need to exist before the app starts

To test locally:
- Drop existing `ratings` database
- Run the first 3 lines of `sql/bootstrap/roles.up.sql`
- Connect to ratings db
- Run the final two lines of the `sql/bootstrap/roles.up.sql`
- Append the `migrations_user` to your `.env` file under `APP_MIGRATION_POSTGRES_URI`
- Make sure the `migrations_user` aligns with the user in your .env file under `APP_MIGRATION_POSTGRES_URI`
- Spin up a gRPC instance with `cargo run`
- In a second terminal, run `cargo test`

We might want to look at moving when we load infrastructure and connect as the `service` db user. As it stands, the `service` db user is needed before the app starts, so can't wait for `migrator.rs` to run to create it. Easiest thing might be to not evaluate the future for the service user connection until it is needed, delaying it to after the migrations have run?

Clippy does complain about Migrator and its functions never being used even thought they are used in the testing module, so ignoring that for now.

Once we finalize these changes I'll append to the HACKING.md how to set up and run the tests locally :slightly_smiling_face:
